commons/src/main/java/org/infinispan/commons/api/functional/FunctionalMap.java

```diff
@@ -0,0 +1,634 @@
+package org.infinispan.commons.api.functional;
```

Comment:
The benefit of having it in commons is that we could potentially provide a remote implementation for it, but I've not made up my mind on that. If the interface is defined in core, we can't have the client access that.